protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+this is a good place to that information. -->
+
+As outlined above, there are different areas where the stack is expanding,
+and where testing needs to expand with it.
+
+What we currently have may be sufficient for a while,
+but the pressure of a "test ceiling" does also stall development.
+E.g. more `op-e2e` bloat -> more `op-e2e` flakes -> less confidence in `op-e2e` -> fewer features / changes.
+
+*Some pressure* may be good: complexity has to stop somewhere.
+But ideally this comes as a design-choice, not a pain/delay in development.
+
+The sections below review what we have today, what challenges we have,
+and what ideas we have, for each of the growth domains.
+
+### Growth of chains
+
+With a greater number of chains, we will need more test deployments,
+and more monitoring.
+
+Today:
+- Superchain-registry [validation checks](https://github.com/ethereum-optimism/superchain-registry/tree/main/validation)
+- Test nodes on each network, with alerts
+- [pessimism](https://github.com/base-org/pessimism) monitoring (deprecated)
+- [monitorism](https://github.com/ethereum-optimism/monitorism)
+
+Challenges:
+- Test nodes can be unstable: lack of peers can result in missed blocks.
+- Test nodes can elevate costs: not everything has to run on the same tier hardware as sequencers.
+  There is a economies-of-scale thing to investigate.
+- Monitorism may need to be deployed to more chains.
+
+
+### Growth of features
+
+"Mo' features, mo' problems." Or really, more edge-cases to validate.
+We need the test-infrastructure to efficiently express how things can go wrong,
+and what we expect to happen in each case.
+
+Today:
+- [`op-e2e` system tests](https://github.com/ethereum-optimism/optimism/tree/develop/op-e2e/system)
+- [`op-e2e` action tests](https://github.com/ethereum-optimism/optimism/tree/develop/op-e2e/actions)
+- [smart contracts tests](https://github.com/ethereum-optimism/optimism/tree/develop/packages/contracts-bedrock/test)
+- [op-chain-ops upgrade checks](https://github.com/ethereum-optimism/optimism/tree/develop/op-chain-ops/cmd)
+- [superchain-ops](https://github.com/ethereum-optimism/superchain-ops/) checks
+- Interop tests (see [interop section](#interoperability))
+
+Challenges:
+- System tests are prone to flakes: running N systems,
+  in a resource-constrained environment, with concurrent work / timers / etc., tends to miss assertions due to processing stalls.
+- `op-e2e` tests are not portable to alternative client implementations.
+  We need ways of generalizing test assertions, and decouple the nodes setup from the test itself.
+- `op-e2e` system tests are quite monolithic. Adding/removing nodes is high friction.
+  Some tests run services that do not influence the test.
+- `op-chain-ops` and `superchain-ops` are not automated against scheduled CI runs.
+  Upgrade checks should run more regularly than just one-off in devnet/testnet.
+- Limited resource sharing: a lot of the feature-tests spawn a copy of the system,
```

Comment:
+1 on this - re-using the L1 (for example) would help with a lot of the flakes we're seeing now.